api/types/provisioning.go

```diff
@@ -302,46 +321,79 @@ func (p ProvisionTokenV2) String() string {
 	return fmt.Sprintf("ProvisionToken(Roles=%v, Expires=%v)", p.Spec.Roles, expires)
 }
 
-// ProvisionTokensToV1 converts provision tokens to V1 list
-func ProvisionTokensToV1(in []ProvisionToken) []ProvisionTokenV1 {
-	if in == nil {
-		return nil
+// V3 returns V3 version of the ProvisionTokenV2 resource.
+func (p *ProvisionTokenV2) V3() *ProvisionTokenV3 {
+	v3 := &ProvisionTokenV3{
+		Kind:     KindToken,
+		Version:  V3,
+		Metadata: p.GetMetadata(),
+		Spec: ProvisionTokenSpecV3{
+			Roles:           p.Spec.Roles,
+			JoinMethod:      p.Spec.JoinMethod,
+			BotName:         p.Spec.BotName,
+			SuggestedLabels: p.Spec.SuggestedLabels,
+		},
 	}
-	out := make([]ProvisionTokenV1, len(in))
-	for i := range in {
-		out[i] = *in[i].V1()
+	// Add provider specific config.
+	switch p.Spec.JoinMethod {
+	case JoinMethodIAM:
+		iam := &ProvisionTokenSpecV3AWSIAM{}
+		for _, rule := range p.Spec.Allow {
+			iam.Allow = append(iam.Allow, &ProvisionTokenSpecV3AWSIAM_Rule{
+				Account: rule.AWSAccount,
+				ARN:     rule.AWSARN,
+			})
+		}
+		v3.Spec.IAM = iam
+	case JoinMethodEC2:
+		ec2 := &ProvisionTokenSpecV3AWSEC2{}
+		for _, rule := range p.Spec.Allow {
+			ec2.Allow = append(ec2.Allow, &ProvisionTokenSpecV3AWSEC2_Rule{
+				Account: rule.AWSAccount,
+				Role:    rule.AWSRole,
+				Regions: rule.AWSRegions,
+			})
+		}
+		ec2.IIDTTL = p.Spec.AWSIIDTTL
+		v3.Spec.EC2 = ec2
```

Comment:
I'd add a default that returns an error when the join method is not supported. Soon we will add Azure to the list. The current version will just skip any new method without any warnings/errors.